docs/tap_hold.md

```diff
@@ -425,6 +425,167 @@ uint16_t get_quick_tap_term(uint16_t keycode, keyrecord_t *record) {
+tap-hold key is pressed and then, before the tapping term, another key is
+pressed. With Chordal Hold, the tap-hold key is settled as tapped if the two
+keys are on the same hand. This behavior may be useful to avoid accidental
+modifier activation with mod-taps, particularly in rolled keypresses when using
+home row mods.
+
+Notes:
+
+* In the case that the keys are on opposite hands, Chordal Hold alone does not
+  yet settle the tap-hold key. Chordal Hold may be used in combination with Hold
+  On Other Key Press or Permissive Hold to determine the behavior. With Hold On
+  Other Key Press, an opposite hands chord is settled immediately as held. Or
+  with Permissive Hold, an opposite hands chord is settled as held provided the
+  other key is pressed and released (nested press) before releasing the tap-hold
+  key.
+
+* Chordal Hold has no effect after the tapping term. 
+
+* Chordal Hold has no effect when the other key is also a tap-hold key. This is
+  so that multiple tap-hold keys may be held on the same hand, which is common
+  to do with home row mods.
+
+* Combos are exempt from the opposite hands rule, since "handedness" is
+  ill-defined in this case. Even so, Chordal Hold's behavior involving combos
+  may be customized through the `get_chordal_hold()` callback.
+
+An example of a sequence that is affected by “chordal hold”: 
+
+- `SFT_T(KC_A)` Down
+- `KC_C` Down
+- `KC_C` Up
+- `SFT_T(KC_A)` Up
+
+```
+                         TAPPING_TERM   
+  +---------------------------|--------+
+  | +----------------------+  |        |
+  | | SFT_T(KC_A)          |  |        |
+  | +----------------------+  |        |
+  |    +--------------+       |        |
+  |    | KC_C         |       |        |
+  |    +--------------+       |        |
+  +---------------------------|--------+
+```
+
+If the two keys are on the same hand, then this will produce `ac` with 
+`SFT_T(KC_A)` settled as tapped the moment that `KC_C` is pressed. 
+
+If the two keys are on opposite hands and the `HOLD_ON_OTHER_KEY_PRESS`
+option enabled, this will produce `C` with `SFT_T(KC_A)` settled as held the
+moment that `KC_C` is pressed.
+
+Or if the two keys are on opposite hands and the `PERMISSIVE_HOLD` option is
+enabled, this will produce `C` with `SFT_T(KC_A)` settled as held the
+moment that `KC_C` is released.
+
+### Chordal Hold Handedness
+
+Determining whether keys are on the same or opposite hands involves defining the
+"handedness" of each key position. By default, if nothing is specified,
+handedness is guessed based on keyboard matrix dimensions. If this is
+inaccurate, handedness may be specified in several ways.
+
+The easiest way to specify handedness is by `chordal_hold_layout`. Define in
+config.h:
+
+```c
+#define CHORDAL_HOLD_LAYOUT
+```
+
+Then in keymap.c, define `chordal_hold_layout` in the following form:
+
+```c
+const char chordal_hold_layout[MATRIX_ROWS][MATRIX_COLS] PROGMEM =
```

Comment:
This is something that could probably be generated automatically based on the matrix config, for most of the dual controller splits.  And doing so may be a good idea.  It can be set to be weak so users can override it, if need be.

Reply:
Excellent idea! Handedness could be inferred from the `.x`, `.y` fields in the `layouts` array in keyboard.json, something like that? I also like that idea to make such a default `chordal_hold_layout` as a weak symbol for easy overriding.

It would be fantastic if handedness could be correctly defined in more cases out of the box. As it is, if no configuration is made, handedness is assumed according to this fallback heuristic based on matrix dimensions:

```c
#if defined(SPLIT_KEYBOARD) || ((MATRIX_ROWS) > (MATRIX_COLS))
// If the keyboard is split or if MATRIX_ROWS > MATRIX_COLS, assume that the
// first half of the rows are left and the latter half are right.
return (key.row < (MATRIX_ROWS) / 2) ? /*left*/ 'L' : /*right*/ 'R';
#else
// Otherwise, assume the first half of the cols are left, others are right.
return (key.col < (MATRIX_COLS) / 2) ? /*left*/ 'L' : /*right*/ 'R';
#endif
```

This simple definition works as expected for my split keyboards. But surely this is not enough in all cases, since there are keyboards with weird matrix circuits.

Reply:
I'm partway on implementing this idea. Here's an update on what I've fleshed out so far...

I've found that the `LAYOUT` macro is generated from keyboard.json by [lib/python/qmk/cli/generate/keyboard_h.py](https://github.com/qmk/qmk_firmware/blob/master/lib/python/qmk/cli/generate/keyboard_h.py). With some simple logic, I can test for the easy case that the keyboard is horizontally symmetric and generate the handedness by checking whether each key's `.x` field lies left or right of the midline. That handles many split keyboards (but not all, e.g. not Charybdis) and some Planck-like keyboards.

In order to make this chordal_hold_layout as a weak symbol variable, I think I need to prepend `__attribute__((weak))` to the beginning of the line, and I believe it's necessary to put the definition in its own .c file (or can a weak variable be overridden by a strong definition in the same translation unit?). I'm planning to do parallel to what you did for Autocorrect, making a new cli tool to generate the code and adding a build rule to invoke it.

An example output for the Moonlander:

```c
const char chordal_hold_layout[MATRIX_ROWS][MATRIX_COLS] PROGMEM = LAYOUT(
    'L', 'L', 'L', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R', 'R', 'R',
    'L', 'L', 'L', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R', 'R', 'R',
    'L', 'L', 'L', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R', 'R', 'R',
    'L', 'L', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R', 'R',
    'L', 'L', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R', 'R',
    'L', 'L', 'L', 'R', 'R', 'R'
);
```

This output is as intended, but the formatting is ugly. If by hand, I would make it like this:

```c
const char chordal_hold_layout[MATRIX_ROWS][MATRIX_COLS] PROGMEM = LAYOUT(
    'L', 'L', 'L', 'L', 'L', 'L', 'L',   'R', 'R', 'R', 'R', 'R', 'R', 'R',
    'L', 'L', 'L', 'L', 'L', 'L', 'L',   'R', 'R', 'R', 'R', 'R', 'R', 'R',
    'L', 'L', 'L', 'L', 'L', 'L', 'L',   'R', 'R', 'R', 'R', 'R', 'R', 'R',
    'L', 'L', 'L', 'L', 'L', 'L',             'R', 'R', 'R', 'R', 'R', 'R',
    'L', 'L', 'L', 'L', 'L',      'L',   'R',      'R', 'R', 'R', 'R', 'R',
                        'L', 'L', 'L',   'R', 'R', 'R'
);
```

I suppose formatting doesn't matter, if this generated code is only an intermediate of the build system and not seen by humans. Or would pretty formatting be worthwhile? Is there an easy way to do it? Thanks again for the idea.

Reply:
I revised handedness configuration: lib/python/qmk/cli/generate/keyboard_c.py generates a default `chordal_hold_layout` array as a weak symbol. Handedness of a key is guessed based on its center x coord relative to the median x value. This heuristic isn't accurate for everything, but it works at least for symmetric split keyboards.

For keyboard makers, a key's handedness can be supplied explicitly in keyboard.json with a new `"hand"` field, like

```json
{"matrix": [5, 6], "x": 0, "y": 5.5, "w": 1.25, "hand", "*"},
```

At the user level, a strong definition `chordal_hold_layout` can be made in keymap.c to override the default definition.